#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [110]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [111]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv(r"C:\Users\DEEPGAMING\Documents\Machineleraning\train2.csv")
train= train.drop(['Unnamed: 0'], axis=1)
train

laptop_ID  Company  TypeName  Inches  ScreenResolution  Ram  OpSys  \
0         1104        5         1    15.6                 2    4      1   
1          172        5         1    17.3                 1    8      1   
2         1150        1         1    15.6                 1   16      1   
3           63        1         1    14.0                 1    8      2   
4          108        3         1    14.0                 1    6      1   
..         ...      ...       ...     ...               ...  ...    ...   
696        123        5         1    15.6                 1    6      1   
697        996        1         1    15.6                 1    6      1   
698        276        4         3    17.3                 1   16      1   
699        585        6         3    17.3                 1   16      1   
700       1173        3         1    15.6                 1    4      1   

     Weight  Price_euros     HDD    SSD  
0      2.40       387.00   500.0    0.0  
1      3.00       854.00     0.0  256.0  
2      2.50      1099.00     0.0    0.0  
3      1.70       599.00     0.0  256.0  
4      1.58       659.00     0.0  256.0  
..      ...          ...     ...    ...  
696    2.10       479.00  1000.0    0.0  
697    2.20       549.00     0.0  256.0  
698    3.20      1549.00  1000.0  256.0  
699    4.14      2729.00  1000.0  512.0  
700    1.91       478.89     0.0  256.0  

[701 rows x 11 columns]

## Data exploration

In [112]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 701 entries, 0 to 700
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         701 non-null    int64  
 1   Company           701 non-null    int64  
 2   TypeName          701 non-null    int64  
 3   Inches            701 non-null    float64
 4   ScreenResolution  701 non-null    int64  
 5   Ram               701 non-null    int64  
 6   OpSys             701 non-null    int64  
 7   Weight            701 non-null    float64
 8   Price_euros       701 non-null    float64
 9   HDD               701 non-null    float64
 10  SSD               701 non-null    float64
dtypes: float64(5), int64(6)
memory usage: 60.4 KB


In [113]:
train.head()

laptop_ID  Company  TypeName  Inches  ScreenResolution  Ram  OpSys  Weight  \
0       1104        5         1    15.6                 2    4      1    2.40   
1        172        5         1    17.3                 1    8      1    3.00   
2       1150        1         1    15.6                 1   16      1    2.50   
3         63        1         1    14.0                 1    8      2    1.70   
4        108        3         1    14.0                 1    6      1    1.58   

   Price_euros    HDD    SSD  
0        387.0  500.0    0.0  
1        854.0    0.0  256.0  
2       1099.0    0.0    0.0  
3        599.0    0.0  256.0  
4        659.0    0.0  256.0

## Data processing

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [114]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(701, 10)

In [115]:
y.shape

(701,)

### 2. Dividir X_train, X_test, y_train, y_test

In [116]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [117]:
X_train

laptop_ID  Company  TypeName  Inches  ScreenResolution  Ram  OpSys  \
165        100        4         1    15.6                 1    6      1   
542       1194        1         1    15.6                 1    4      4   
284        883        7         1    15.6                 1    8      1   
360        935        3         2    15.6                 1    8      1   
110       1093        1         1    15.6                 2    4      1   
..         ...      ...       ...     ...               ...  ...    ...   
71         673        7         1    15.6                 2    4      1   
106       1318        1         1    14.0                 2    2      1   
270        710        1         1    15.6                 2    8      1   
435        909        3         1    15.6                 1    8      1   
102        739        2         1    15.6                 2   12      1   

     Weight     HDD    SSD  
165    2.00     0.0  256.0  
542    2.32     0.0    0.0  
284    2.00     0.0  256.0  
360    1.84     0.0  256.0  
110    2.10   500.0    0.0  
..      ...     ...    ...  
71     2.10     0.0  128.0  
106    1.50     0.0    0.0  
270    2.20  1000.0    0.0  
435    2.04  1000.0    0.0  
102    2.25  1000.0    0.0  

[560 rows x 10 columns]

### 3. Asignar el modelo (vacío) a una variable


In [118]:
from sklearn.linear_model import LinearRegression  

model = LinearRegression()

In [119]:
model.fit((X_train), (y_train))

LinearRegression()

In [120]:
predict = model.predict((X_test))
predict

array([1279.75771092, 1171.10835514,  827.28204976,  625.70407404,
       1214.28898671,  358.07829627,  999.78054895,  631.6164544 ,
       1270.25637362,  904.55747868,  330.32647336, 1496.15255426,
       2748.92512958,  721.00285321, 1055.21703059, 1082.20093204,
        476.63648515,  682.63672938,  515.34029826, 1034.73819837,
        865.61535093, 1390.85692143, 1288.76499382,  641.30466122,
       1622.77002044, 1305.84000076, 1822.69377647,  414.48044829,
        808.01090057,  374.20801645, 1091.93870881,  556.20634105,
        788.41131935, 1228.54078773,  677.87571766, 1263.25396292,
       1287.73097556,  496.2049214 ,  923.01394916,  776.61147524,
        473.33909234, 1705.02224293, 1195.04983118, 1252.97068303,
       1966.99400304, 1032.06510024,  459.82146477, 1883.24094999,
       1085.57236762, 1126.34891686,  968.81360556,  518.30448652,
       1112.55309894, 1386.11250497, 1223.46993718, 1004.2965541 ,
        554.69612643,  790.69036348, 1150.9442471 ,  634.04042

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [121]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE:',np.sqrt(mean_squared_error(y_test, predict)))

RMSE: 385.05786551207876


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [122]:
from sklearn.linear_model import LinearRegression  

model = LinearRegression()

In [123]:
model.fit((X), (y))

LinearRegression()

### 2. Carga los datos de `test.csv` para predecir.

In [210]:
test=pd.read_csv(r"C:\Users\DEEPGAMING\Documents\Machineleraning\Machine_Learning\1-Supervisado\1-Regression\Kaggle1\data\test.csv")

In [211]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         391 non-null    int64  
 1   Company           391 non-null    object 
 2   Product           391 non-null    object 
 3   TypeName          391 non-null    object 
 4   Inches            391 non-null    float64
 5   ScreenResolution  391 non-null    object 
 6   Cpu               391 non-null    object 
 7   Ram               391 non-null    object 
 8   Memory            391 non-null    object 
 9   Gpu               391 non-null    object 
 10  OpSys             391 non-null    object 
 11  Weight            391 non-null    object 
dtypes: float64(1), int64(1), object(10)
memory usage: 36.8+ KB


In [212]:
company_dict = {'Lenovo':1, 'Dell':2, 'HP':3, 'Asus':4, 'Acer':5, 'MSI':6, 'Toshiba':7,'Xiaomi':8,'Apple':9,'Samsung':10,'Razer':11,'Fujitsu':12,'Microsoft':13,'LG':14,'Mediacom':15,'Google':16,'Vero':17,'Huawei':18,'Chuwi':19}
test['Company'] = test['Company'].replace(company_dict)
test['Company']=test['Company'].astype('int64')
test['Company'].value_counts()

2     85
1     82
3     78
4     54
5     34
6     19
7     16
9      5
11     4
10     4
17     3
15     2
19     2
12     1
18     1
14     1
Name: Company, dtype: int64

In [213]:
TypeName_dict = {'Notebook':1, 'Ultrabook':2, 'Gaming':3, '2 in 1 Convertible':4,'Workstation':5,'Netbook':6}
test['TypeName'] = test['TypeName'].replace(TypeName_dict)
test['TypeName']=test['TypeName'].astype('int64')
test['TypeName'].value_counts()

1    212
3     71
2     55
4     38
5      9
6      6
Name: TypeName, dtype: int64

In [214]:

test['Inches'].value_counts()

15.6    211
17.3     54
13.3     48
14.0     46
11.6     11
12.5     10
10.1      2
13.9      2
12.3      2
15.0      2
15.4      1
11.3      1
13.0      1
Name: Inches, dtype: int64

In [215]:
test['ScreenResolution'].value_counts()

Full HD 1920x1080                                162
IPS Panel Full HD 1920x1080                       72
1366x768                                          69
IPS Panel Full HD / Touchscreen 1920x1080         21
Full HD / Touchscreen 1920x1080                   11
1600x900                                           7
Touchscreen 1366x768                               7
4K Ultra HD / Touchscreen 3840x2160                5
IPS Panel Retina Display 2560x1600                 4
Quad HD+ / Touchscreen 3200x1800                   3
IPS Panel 1366x768                                 3
IPS Panel 4K Ultra HD 3840x2160                    2
4K Ultra HD 3840x2160                              2
IPS Panel Touchscreen 1920x1200                    2
Touchscreen 2560x1440                              2
IPS Panel Quad HD+ / Touchscreen 3200x1800         2
IPS Panel Full HD 1366x768                         1
2560x1440                                          1
IPS Panel Touchscreen 2560x1440               

In [216]:
test['ScreenResolution']=test['ScreenResolution'].str[-10:]
test['ScreenResolution']= test['ScreenResolution'].str.replace('n','')
test['ScreenResolution']= test['ScreenResolution'].str.replace('l','')
test['ScreenResolution']= test['ScreenResolution'].str.replace('D','')
test['ScreenResolution']= test['ScreenResolution'].str.replace(' ','')

test['ScreenResolution'].value_counts()

1920x1080    268
1366x768      81
3840x2160     11
3200x1800      7
1600x900       7
2560x1440      6
2560x1600      4
1920x1200      3
2880x1800      1
2400x1600      1
2160x1440      1
2736x1824      1
Name: ScreenResolution, dtype: int64

In [217]:
Screen_dict = {'1920x1080':1, '1366x768':2,'3840x2160':3, '3200x1800':4, '2560x1440':5,'2560x1600':6 ,'1920x1200':7,'2880x1800':8,'2400x1600':9,'2160x1440':10,'2736x1824':11,'1600x900':12}
test['ScreenResolution'] = test['ScreenResolution'].replace(Screen_dict)
test['ScreenResolution'].value_counts()

1     268
2      81
3      11
4       7
12      7
5       6
6       4
7       3
8       1
9       1
10      1
11      1
Name: ScreenResolution, dtype: int64

In [204]:
test["ScreenResolution"].astype('float64')
test["ScreenResolution"].value_counts()

1     268
2      81
3      11
4       7
12      7
5       6
6       4
7       3
8       1
9       1
10      1
11      1
Name: ScreenResolution, dtype: int64

In [218]:
for index, row in test.iterrows():
    # Comprobar si la descripción del disco duro contiene "SSD"
    if "SSD" in row["Memory"]:
        test.loc[index, "SSD"] = row["Memory"]
    # Comprobar si la descripción del disco duro contiene "HDD"
    if "HDD" in row["Memory"]:
        test.loc[index, "HDD"] = row["Memory"]

In [219]:
test['SSD']=test['SSD'].str[:5]
test['SSD']=test['SSD'].str.replace('+','')
test['SSD']=test['SSD'].str.replace('GB','')
test['SSD']=test['SSD'].str.replace('S','')
test['SSD']=test['SSD'].str.replace('TB','000')
#test['SSD']=test['SSD'].str.replace(' ','')
test['SSD']=test['SSD'].fillna(0)
test['SSD']

C:\Users\DEEPGAMING\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  


0      256
1        0
2        0
3        0
4        0
      ... 
386    256
387    256
388    256
389      0
390    240
Name: SSD, Length: 391, dtype: object

In [220]:
test['SSD']=test['SSD'].astype("float64")
test['SSD']

0      256.0
1        0.0
2        0.0
3        0.0
4        0.0
       ...  
386    256.0
387    256.0
388    256.0
389      0.0
390    240.0
Name: SSD, Length: 391, dtype: float64

In [221]:
test['HDD']=test['HDD'].str[-10:]
test['HDD']=test['HDD'].str.replace('GB','')
test['HDD']=test['HDD'].str.replace('TB','000')
test['HDD']=test['HDD'].str.replace('+','')
test['HDD']=test['HDD'].str.replace(' ','')
test['HDD']=test['HDD'].str.replace('HDD','')
test['HDD']=test['HDD'].fillna(0)
test['HDD']


C:\Users\DEEPGAMING\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  after removing the cwd from sys.path.


0         0
1        32
2         0
3      1000
4         0
       ... 
386    1000
387       0
388       0
389    1000
390       0
Name: HDD, Length: 391, dtype: object

In [222]:
test['HDD']=test['HDD'].astype("float64")

In [223]:
test["Ram"]=test['Ram'].str.replace('GB','')
test["Ram"]=test["Ram"].astype("int64")
test["Ram"].value_counts()

8     189
4     102
16     62
6      14
2       8
12      8
32      5
24      2
64      1
Name: Ram, dtype: int64

In [ ]:
test['OpSys'].value_counts()

Windows 10      321
Linux            22
No OS            19
Windows 7        14
Chrome OS         8
macOS             5
Android           1
Windows 10 S      1
Name: OpSys, dtype: int64

In [224]:

OpSys_dict = {'Windows 10':1, 'No OS':2, 'Linux':3, 'Windows 7':4,'Chrome OS':5,'Windows 10 S':6,'macOS':7,'Android':8}
test['OpSys'] = test['OpSys'].replace(OpSys_dict)
test['OpSys']=test['OpSys'].astype('int64')
test['OpSys'].value_counts()

1    321
3     22
2     19
4     14
5      8
7      5
8      1
6      1
Name: OpSys, dtype: int64

In [225]:
test["Weight"]=test['Weight'].str.replace('kg','')
test["Weight"]=test["Weight"].astype("float64")
test["Weight"].value_counts()  

2.20    34
2.40    18
2.10    17
2.30    16
2.00    14
        ..
2.15     1
1.85     1
1.35     1
1.44     1
1.39     1
Name: Weight, Length: 116, dtype: int64

In [226]:
test = test.drop(['Product','Cpu','Memory',"Gpu"],axis=1 )

In [227]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         391 non-null    int64  
 1   Company           391 non-null    int64  
 2   TypeName          391 non-null    int64  
 3   Inches            391 non-null    float64
 4   ScreenResolution  391 non-null    int64  
 5   Ram               391 non-null    int64  
 6   OpSys             391 non-null    int64  
 7   Weight            391 non-null    float64
 8   SSD               391 non-null    float64
 9   HDD               391 non-null    float64
dtypes: float64(4), int64(6)
memory usage: 30.7 KB


In [228]:
X_pred = test
X_pred.head()

laptop_ID  Company  TypeName  Inches  ScreenResolution  Ram  OpSys  Weight  \
0        750        7         1    14.0                 1    8      1    1.25   
1        726       15         1    14.0                 7    4      1    1.40   
2        633       15         1    14.0                 1    2      1    1.40   
3        363        2         3    15.6                 1    8      1    2.65   
4        319        4         4    11.6                 2    2      1    1.10   

     SSD     HDD  
0  256.0     0.0  
1    0.0    32.0  
2    0.0     0.0  
3    0.0  1000.0  
4    0.0     0.0

In [229]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         391 non-null    int64  
 1   Company           391 non-null    int64  
 2   TypeName          391 non-null    int64  
 3   Inches            391 non-null    float64
 4   ScreenResolution  391 non-null    int64  
 5   Ram               391 non-null    int64  
 6   OpSys             391 non-null    int64  
 7   Weight            391 non-null    float64
 8   SSD               391 non-null    float64
 9   HDD               391 non-null    float64
dtypes: float64(4), int64(6)
memory usage: 30.7 KB


In [230]:
X_pred.shape

(391, 10)

**¿Por qué da error?** 

In [231]:
predictions_submit = model.predict(X_pred)
predictions_submit

c:\Users\DEEPGAMING\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


array([  971.88813803,  -106.56257992,   881.33731969,  1920.04698899,
         627.91652898,   498.40982313,  1374.70750952,  2500.98264868,
        1413.87387301,  1975.03972293,   699.88275849,   813.41906011,
        1964.90052784,   533.4633915 ,   494.835858  ,   826.71886438,
         824.6503909 ,  1768.36647362,  1608.27212218,  1776.83288426,
         889.12044711,  1711.08537492,   992.74942232,   721.71296579,
         423.86496776,    83.04103739,  1598.44598345,  2575.81616511,
        1534.33669827,  1936.9119335 ,  1645.04813362,  1777.34829649,
         766.28346007,  1961.44735951,  1165.75630991,   946.4980789 ,
        1752.2442415 ,  1056.65134844,  2596.63163073,  1548.95330883,
        2620.52407619,  1814.14702529,   349.30382386,    59.64908674,
         744.1409115 ,  2578.87301585,  1909.98254349,  2398.38893387,
         871.36073443,  1562.5795979 ,  1172.09258796,  1474.3624249 ,
        1704.77082515,   883.40924586,   275.03042606,   651.36155168,
      

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [232]:
predictions_submit = model.predict(X_pred)
predictions_submit

c:\Users\DEEPGAMING\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


array([  971.88813803,  -106.56257992,   881.33731969,  1920.04698899,
         627.91652898,   498.40982313,  1374.70750952,  2500.98264868,
        1413.87387301,  1975.03972293,   699.88275849,   813.41906011,
        1964.90052784,   533.4633915 ,   494.835858  ,   826.71886438,
         824.6503909 ,  1768.36647362,  1608.27212218,  1776.83288426,
         889.12044711,  1711.08537492,   992.74942232,   721.71296579,
         423.86496776,    83.04103739,  1598.44598345,  2575.81616511,
        1534.33669827,  1936.9119335 ,  1645.04813362,  1777.34829649,
         766.28346007,  1961.44735951,  1165.75630991,   946.4980789 ,
        1752.2442415 ,  1056.65134844,  2596.63163073,  1548.95330883,
        2620.52407619,  1814.14702529,   349.30382386,    59.64908674,
         744.1409115 ,  2578.87301585,  1909.98254349,  2398.38893387,
         871.36073443,  1562.5795979 ,  1172.09258796,  1474.3624249 ,
        1704.77082515,   883.40924586,   275.03042606,   651.36155168,
      

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [233]:
sample = pd.read_csv(r"C:\Users\DEEPGAMING\Documents\Machineleraning\Machine_Learning\1-Supervisado\1-Regression\Kaggle1\data\sample_submission.csv")
sample

laptop_ID  Price_euros
0          750          500
1          726          500
2          633          500
3          363          650
4          319          650
..         ...          ...
386        693          650
387       1132          500
388         49          650
389       1243          650
390        924          500

[391 rows x 2 columns]

In [234]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [235]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [257]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [258]:
submission.head()

laptop_ID  Price_euros
0        750   971.888138
1        726  -106.562580
2        633   881.337320
3        363  1920.046989
4        319   627.916529

In [259]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [260]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [261]:
chequeator(submission)

You're ready to submit!
